### PR TITLE
DSM Demo and Documentation

### PR DESCRIPTION
**Pull request by PolygonalSun**
This PR contains a link to a demo for the devicesourcemanager and its corresponding documentation.

### CODE REVIEW

**Comment by Drigax** [comment]
Do you have a link to the sample playground in your documentation page?

File: content/How_To/events/How_to_use_DeviceSourceManager.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
+---
+PG_TITLE: How to use the DeviceSourceManager
+---
+
+# DeviceSourceManager
+The [DeviceSourceManager](https://doc.babylonjs.com/api/classes/babylon.devicesourcemanager) is a class that will manage the connections for various user input devices and provide methods of querying those devices for their current state.
```
Comment (Drigax) [inline content/How_To/events/How_to_use_DeviceSourceManager.md]: Maybe also expand a bit on the scope of the supported input devices? what types of devices can a user reasonably expect to work seamlessly using this device source manager?

Comment (PolygonalSun) [reply content/How_To/events/How_to_use_DeviceSourceManager.md]: Good call, I'll get that added.

File: content/How_To/events/How_to_use_DeviceSourceManager.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+To use this, first create a DeviceSourceManager.  You will need to provide an engine object.
+```javascript
+var deviceSourceManager = new BABYLON.deviceSourceManager(scene.getEngine());
```
Comment (bghgary) [inline content/How_To/events/How_to_use_DeviceSourceManager.md]:
```suggestion
var deviceSourceManager = new BABYLON.DeviceSourceManager(scene.getEngine());
```

File: content/How_To/events/How_to_use_DeviceSourceManager.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+```
+
+Within your scene's render/game loop, you can query the DeviceSourceManager for the current state of a specific input.  First, you will need to get the DeviceSource object.  With this object, you can then query for a specific input's status.
+
+In Javascript, you should use the following approach
+```javascript
+// If the device has been registered in the DSM
+if (deviceSourceManager.getDeviceSource(BABYLON.DeviceType.Xbox)) {
+    // And the A button was pressed
+    if (deviceSourceManager.getDeviceSource(BABYLON.DeviceType.Xbox).getInput(BABYLON.XboxInput.A) ==  1) {
```
Comment (bghgary) [inline content/How_To/events/How_to_use_DeviceSourceManager.md]: nit:
```suggestion
    if (deviceSourceManager.getDeviceSource(BABYLON.DeviceType.Xbox).getInput(BABYLON.XboxInput.A) == 1) {
```

File: content/How_To/events/How_to_use_DeviceSourceManager.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+To use this, first create a DeviceSourceManager.  You will need to provide an engine object.
+```javascript
+var deviceSourceManager = new BABYLON.deviceSourceManager(scene.getEngine());
+```
+
+Within your scene's render/game loop, you can query the DeviceSourceManager for the current state of a specific input.  First, you will need to get the DeviceSource object.  With this object, you can then query for a specific input's status.
+
+In Javascript, you should use the following approach
+```javascript
+// If the device has been registered in the DSM
```
Comment (bghgary) [inline content/How_To/events/How_to_use_DeviceSourceManager.md]: I would recommend spelling everything out in documentation unless you explicitly say what the initialism means.
```suggestion
// If the device has been registered in the device source manager
```

File: content/How_To/events/How_to_use_DeviceSourceManager.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    }
+}
+```
+
+In Typescript, you can use a null-conditional operator to make it a one line call
+```javascript
+if (deviceSourceManager.getDeviceSource(BABYLON.DeviceType.Xbox)?.getInput(BABYLON.XboxInput.A) ==  1) {
+    // Do something
+}
+```
```
Comment (bghgary) [inline content/How_To/events/How_to_use_DeviceSourceManager.md]: If you are using TypeScript, it's probable you are also using modules, so you probably don't have the `BABYLON` prefix.

File: content/How_To/events/How_to_use_DeviceSourceManager.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+# Events and Observables
+You can use the following Observables to work with identifiers for a given device
+```javascript
+/*  Before a device is registered */
```
Comment (bghgary) [inline content/How_To/events/How_to_use_DeviceSourceManager.md]: nit: might consider using `//` for consistency

File: examples/list.json [hunk 7]
```diff
@@ -150,6 +150,13 @@
...
                     "description": "Position, Rotate and Scale meshes with a pointer",
                     "PGID": "#8MGKWK#25"
+                },
+				{
+                    "title": "Using DeviceSourceManager",
+                    "doc": "https://doc.babylonjs.com/how_to/how_to_devicesourcemanager",
+                    "icon": "icons/dsm_input.png",
+                    "description": "Control ship with keyboard or gamepad.",
+                    "PGID": "#C7PM2B"
                 }
```
Comment (bghgary) [inline examples/list.json]:
```suggestion
                {
                    "title": "Using DeviceSourceManager",
                    "doc": "https://doc.babylonjs.com/how_to/how_to_devicesourcemanager",
                    "icon": "icons/dsm_input.png",
                    "description": "Control ship with keyboard or gamepad.",
                    "PGID": "#C7PM2B"
                }
```

File: content/How_To/events/How_to_use_DeviceSourceManager.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+
+To use this, first create a DeviceSourceManager.  You will need to provide an engine object.
+```javascript
+var deviceSourceManager = new BABYLON.deviceSourceManager(scene.getEngine());
+```
+
+Within your scene's render/game loop, you can query the DeviceSourceManager for the current state of a specific input.  First, you will need to get the DeviceSource object.  With this object, you can then query for a specific input's status.
+
+In Javascript, you should use the following approach
```
Comment (bghgary) [inline content/How_To/events/How_to_use_DeviceSourceManager.md]: [Optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining#Browser_compatibility) is mostly already available in major browsers for JavaScript, so I would just use it without saying anything about TypeScript. Maybe you can mention "optional chaining" as a way to shorten the line.

File: content/How_To/events/How_to_use_DeviceSourceManager.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+For Keyboards and Pointers, you can use an event based system to get the current input and previous input when an input is activated
+```javascript
+deviceSourceManager.getDeviceSource(BABYLON.DeviceType.Keyboard).onInputChangedObservable.add((device) => {
+    // device.inputIndex is the activated input
+    // device.currentState is the current value
+    // device.previousState is the previous value (before activation)
+});
+```
+* [Playground Example](https://playground.babylonjs.com/#C7PM2B)
```
Comment (bghgary) [inline content/How_To/events/How_to_use_DeviceSourceManager.md]: Maybe it's better to put this up top?

File: content/How_To/events/How_to_use_DeviceSourceManager.md (hunk 8 above)
Comment (PolygonalSun) [reply content/How_To/events/How_to_use_DeviceSourceManager.md]: Ah, okay.  I wasn't sure if got added yet.  Is there any reason why optional chaining doesn't work in the PG?

File: content/How_To/events/How_to_use_DeviceSourceManager.md (hunk 9 above)
Comment (PolygonalSun) [reply content/How_To/events/How_to_use_DeviceSourceManager.md]: Moved

File: content/How_To/events/How_to_use_DeviceSourceManager.md (hunk 8 above)
Comment (PolygonalSun) [reply content/How_To/events/How_to_use_DeviceSourceManager.md]: Changed section

File: examples/list.json (hunk 7 above)
Comment (PolygonalSun) [reply examples/list.json]: Fixed spacing

File: content/How_To/events/How_to_use_DeviceSourceManager.md (hunk 6 above)
Comment (PolygonalSun) [reply content/How_To/events/How_to_use_DeviceSourceManager.md]: Changed

File: content/How_To/events/How_to_use_DeviceSourceManager.md (hunk 5 above)
Comment (PolygonalSun) [reply content/How_To/events/How_to_use_DeviceSourceManager.md]: I removed the explicit ref to Typescript

File: content/How_To/events/How_to_use_DeviceSourceManager.md (hunk 4 above)
Comment (PolygonalSun) [reply content/How_To/events/How_to_use_DeviceSourceManager.md]: Changed

**Review by bghgary — APPROVED** [review]

File: content/How_To/events/How_to_use_DeviceSourceManager.md (hunk 3 above)
Comment (PolygonalSun) [reply content/How_To/events/How_to_use_DeviceSourceManager.md]: fixed

File: content/How_To/events/How_to_use_DeviceSourceManager.md (hunk 2 above)
Comment (PolygonalSun) [reply content/How_To/events/How_to_use_DeviceSourceManager.md]: fixed

File: content/How_To/events/How_to_use_DeviceSourceManager.md (hunk 1 above)
Comment (PolygonalSun) [reply content/How_To/events/How_to_use_DeviceSourceManager.md]: Expanded section with supported devices

File: content/How_To/events/How_to_use_DeviceSourceManager.md (hunk 8 above)
Comment (bghgary) [reply content/How_To/events/How_to_use_DeviceSourceManager.md]: It should work? Does it not work?

Comment (PolygonalSun) [reply content/How_To/events/How_to_use_DeviceSourceManager.md]: Nope, I've tried it and it doesn't work

Comment (bghgary) [reply content/How_To/events/How_to_use_DeviceSourceManager.md]: Oh, it looks like the feature is behind a flag.

> From version 79: this feature is behind the Experimental JavaScript preference (needs to be set to true).

**Pull request merged**